File: Estruturas de Dados e Complexidade de Algoritmos/Algoritmos15/Alg15.rb
=begin
♦ Algoritmo 15: Estrutura de Dados Linear Dinâmica (não Estática) de Pilha: Stack (Pilha).
=end

class PilhaDinamica
  def initialize
    @pilha = []
  end

  def is_empty
    @pilha.empty?
  end

  def push(elemento)
    @pilha.unshift(elemento)
  end

  def pop
    if is_empty
      raise "A pilha está vazia"
    end
    @pilha.shift
  end

  def size
    @pilha.length
  end

  def front
    if is_empty
      raise "A pilha está vazia"
    end
    @pilha.first
  end

  def show_stack
    puts @pilha.join(" ")
  end
end

pilha = PilhaDinamica.new

pilha.push(1)
pilha.push(2)
pilha.push(3)

pilha.show_stack

puts "Tamanho da pilha: #{pilha.size}"
puts "Topo da pilha: #{pilha.front}"

pilha.pop
pilha.show_stack

File: Estruturas de Dados e Complexidade de Algoritmos/Algoritmos15/Alg15.java
/*
♦ Algoritmo 15: Estrutura de Dados Linear Dinâmica (não Estática) de Pilha: Stack (Pilha).
*/

import java.util.LinkedList;

public class Alg15<T> {
    private LinkedList<T> pilha;

    public Alg15() {
        pilha = new LinkedList<>();
    }

    public boolean isEmpty() {
        return pilha.isEmpty();
    }

    public void push(T elemento) {
        pilha.addFirst(elemento);
    }

    public void pop() {
        if (isEmpty()) {
            throw new IllegalStateException("A pilha está vazia");
        }
        pilha.removeFirst();
    }

    public int size() {
        return pilha.size();
    }

    public T front() {
        if (isEmpty()) {
            throw new IllegalStateException("A pilha está vazia");
        }
        return pilha.getFirst();
    }

    public void showStack() {
        for (T elemento : pilha) {
            System.out.print(elemento + " ");
        }
        System.out.println();
    }

    public static void main(String[] args) {
        Alg15<Integer> pilha = new Alg15<>();

        pilha.push(1);
        pilha.push(2);
        pilha.push(3);

        pilha.showStack();

        System.out.println("Tamanho da pilha: " + pilha.size());
        System.out.println("Topo da pilha: " + pilha.front());

        pilha.pop();
        pilha.showStack();
    }
}
